Greengrass DiscoveryClient does not work if a SecurityManager is installed

### Describe the bug

If a SecurityManager is installed, Greengrass Core Device discovery performed using DiscoveryClient will fail even with a permissive policy.



### Expected Behavior

Greengrass Discovery completes successfully even if a security manager is installed.

### Current Behavior

Greengrass Core Device discovery performed using DiscoveryClient fails with exception [1]

[1]
```
Exception thrown: java.util.concurrent.ExecutionException: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
java.util.concurrent.ExecutionException: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at greengrass.BasicDiscovery.getClientFromDiscovery(BasicDiscovery.java:157)
	at greengrass.BasicDiscovery.main(BasicDiscovery.java:114)
Caused by: com.google.gson.JsonIOException: Failed making field 'software.amazon.awssdk.iot.discovery.model.DiscoverResponse#GGGroups' accessible; either change its visibility or write a custom TypeAdapter for its declaring type
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:22)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:158)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:101)
	at com.google.gson.Gson.getAdapter(Gson.java:501)
	at com.google.gson.Gson.fromJson(Gson.java:990)
	at com.google.gson.Gson.fromJson(Gson.java:929)
	at software.amazon.awssdk.iot.discovery.DiscoveryClient.lambda$discover$0(DiscoveryClient.java:97)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1596)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
Caused by: java.security.AccessControlException: access denied ("java.lang.reflect.ReflectPermission" "suppressAccessChecks")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
	at java.security.AccessController.checkPermission(AccessController.java:886)
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
	at java.lang.reflect.AccessibleObject.setAccessible(AccessibleObject.java:128)
	at com.google.gson.internal.reflect.ReflectionHelper.makeAccessible(ReflectionHelper.java:19)
	... 12 more
```

### Reproduction Steps

Build a standalone JAR file that executes https://github.com/aws/aws-iot-device-sdk-java-v2/blob/main/samples/Greengrass/src/main/java/greengrass/BasicDiscovery.java as main class. This can obtained for example by adding the snippet [1] to the build -> plugins section of https://github.com/aws/aws-iot-device-sdk-java-v2/blob/main/samples/Greengrass/pom.xml.

Run the produced jar file with the following command:

`java -Djava.security.manager -Djava.security.policy=security.policy -jar ./Greengrass-1.0-SNAPSHOT-jar-with-dependencies.jar ${DISCOVERY_SAMPLE_ARGS}`

where `DISCOVERY_SAMPLE_ARGS` is as described in https://docs.aws.amazon.com/greengrass/v2/developerguide/test-client-device-communications.html#test-client-device-communications-java and `security.policy` contains the following maximally permissive policy:

```
grant { 
      permission java.security.AllPermission;
};
```

The command should fail logging the reported exception if the`-Djava.security.manager -Djava.security.policy=security.policy` arguments are provided and succeed if they are removed.

[1]
```
<plugin>
    <artifactId>maven-assembly-plugin</artifactId>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>single</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <archive>
            <manifest>
                <mainClass>greengrass.BasicDiscovery</mainClass>
            </manifest>
        </archive>
        <descriptorRefs>
            <descriptorRef>jar-with-dependencies</descriptorRef>
        </descriptorRefs>
    </configuration>
</plugin>
```


### Possible Solution

The issue is likely caused by the fact that

https://github.com/aws/aws-iot-device-sdk-java-v2/blob/472d0ecd511b3e31ee65882636dc91cbcd5d1264/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java#L97

is executed through [CompletableFuture.supplyAsync()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#supplyAsync-java.util.function.Supplier-) using the Fork Join Pool common pool.

Quoting from [ForkJoinPool Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html), _If a [SecurityManager](https://docs.oracle.com/javase/8/docs/api/java/lang/SecurityManager.html) is present and no factory is specified, then the default pool uses a factory supplying threads that have no [Permissions](https://docs.oracle.com/javase/8/docs/api/java/security/Permissions.html) enabled._.

### Additional Information/Context

_No response_

### SDK version used

main branch cloned from https://github.com/aws/aws-iot-device-sdk-java-v2/commit/1a2e1b5d82eab57e8365ca91883dc1ad9f3a32c3

### Environment details (OS name and version, etc.)

Raspberry PI OS, openjdk version "1.8.0_312" OpenJDK Runtime Environment (build 1.8.0_312-8u312-b07-1+rpi1-b07) OpenJDK Client VM (build 25.312-b07, mixed mode)

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
@@ -12,6 +12,9 @@ import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 /**
  * Class for performing network-based discovery of the connectivity properties of registered greengrass cores
@@ -35,11 +38,26 @@ public class DiscoveryClient implements AutoCloseable {
     private final HttpClientConnectionManager httpClientConnectionManager;
 
     /**
+     * We need to use a code defined executor to avoid leaving threads alive when the discovery client is closed
+     * It also fixes issues with the SecurityManager as well - as created ExecutorServices created via code
+     * inherit the permissions of the application, unlike the default common thread pool that is used by default
+     * with supplyAsync otherwise.
+     */
+    private ExecutorService executorService = null;
+    private boolean cleanExecutor = false;
+
+    /**
      *
      * @param config Greengrass discovery client configuration
      */
     public DiscoveryClient(final DiscoveryClientConfig config) {
-        this.httpClientConnectionManager = HttpClientConnectionManager.create(
+        executorService = config.getDiscoveryExecutor();
+        if (executorService == null) {
+            // If an executor is not set, then create one and make sure to clean it when finished.
+            executorService = Executors.newFixedThreadPool(1);
+            cleanExecutor = true;
+        }
+        httpClientConnectionManager = HttpClientConnectionManager.create(
                 new HttpClientConnectionManagerOptions()
                     .withClientBootstrap(config.getBootstrap())
                     .withProxyOptions(config.getProxyOptions())
@@ -59,57 +77,49 @@ public class DiscoveryClient implements AutoCloseable {
         if(thingName == null) {
             throw new IllegalArgumentException("ThingName cannot be null!");
         }
-
-        CompletableFuture<DiscoverResponse> result = new CompletableFuture<>();
-        // We create and start a thread so it executes asynchronously and independently from the main process. The thread will set the
-        // result CompletableFuture when it is done or if an exception occurs.
-        new Thread(() -> {
-            try {
-                try (final HttpClientConnection connection = httpClientConnectionManager.acquireConnection().get()) {
-                    final String requestHttpPath =  "/greengrass/discover/thing/" + thingName;
-                    final HttpHeader[] headers = new HttpHeader[] {
-                            new HttpHeader("host", httpClientConnectionManager.getUri().getHost())
-                    };
-                    final HttpRequest request = new HttpRequest("GET", requestHttpPath, headers, null);
-                    //we are storing everything until we get the entire response
-                    final CompletableFuture<Integer> responseComplete = new CompletableFuture<>();
-                    final StringBuilder jsonBodyResponseBuilder = new StringBuilder();
-                    final Map<String, String> responseInfo = new HashMap<>();
-
-                    try (final HttpStream stream = connection.makeRequest(request, new HttpStreamResponseHandler() {
-                            @Override
-                            public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] httpHeaders) {
-                                Arrays.stream(httpHeaders).forEach(header -> {
-                                    responseInfo.put(header.getName(), header.getValue());
-                                });
-                            }
-                            @Override
-                            public int onResponseBody(HttpStream stream, byte bodyBytes[]) {
-                                jsonBodyResponseBuilder.append(new String(bodyBytes, StandardCharsets.UTF_8));
-                                return bodyBytes.length;
-                            }
-                            @Override
-                            public void onResponseComplete(HttpStream httpStream, int errorCode) {
-                                responseComplete.complete(errorCode);
-                            }
-                         }))
-                         {
-                            stream.activate();
-                            responseComplete.get();
-                            if (stream.getResponseStatusCode() != 200) {
-                                throw new RuntimeException(String.format("Error %s(%d); RequestId: %s",
-                                        HTTP_HEADER_ERROR_TYPE, stream.getResponseStatusCode(), HTTP_HEADER_REQUEST_ID));
-                            }
-                            final String responseString = jsonBodyResponseBuilder.toString();
-                            result.complete(GSON.fromJson(new StringReader(responseString), DiscoverResponse.class));
+        return CompletableFuture.supplyAsync(() -> {
+            try(final HttpClientConnection connection = httpClientConnectionManager.acquireConnection().get()) {
+                final String requestHttpPath =  "/greengrass/discover/thing/" + thingName;
+                final HttpHeader[] headers = new HttpHeader[] {
+                        new HttpHeader("host", httpClientConnectionManager.getUri().getHost())
+                };
+                final HttpRequest request = new HttpRequest("GET", requestHttpPath, headers, null);
+                //we are storing everything until we get the entire response
+                final CompletableFuture<Integer> responseComplete = new CompletableFuture<>();
+                final StringBuilder jsonBodyResponseBuilder = new StringBuilder();
+                final Map<String, String> responseInfo = new HashMap<>();
+                try(final HttpStream stream = connection.makeRequest(request, new HttpStreamResponseHandler() {
+                        @Override
+                        public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] httpHeaders) {
+                            Arrays.stream(httpHeaders).forEach(header -> {
+                                responseInfo.put(header.getName(), header.getValue());
+                            });
                         }
+                        @Override
+                        public int onResponseBody(HttpStream stream, byte bodyBytes[]) {
+                            jsonBodyResponseBuilder.append(new String(bodyBytes, StandardCharsets.UTF_8));
+                            return bodyBytes.length;
+                        }
+                        @Override
+                        public void onResponseComplete(HttpStream httpStream, int errorCode) {
+                            responseComplete.complete(errorCode);
+                        }})) {
+                    stream.activate();
+                    responseComplete.get();
+                    if (stream.getResponseStatusCode() != 200) {
+                        throw new RuntimeException(String.format("Error %s(%d); RequestId: %s",
+                                HTTP_HEADER_ERROR_TYPE, stream.getResponseStatusCode(), HTTP_HEADER_REQUEST_ID));
+                    }
+                    final String responseString = jsonBodyResponseBuilder.toString();
+                    return GSON.fromJson(new StringReader(responseString), DiscoverResponse.class);
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new RuntimeException(e);
                 }
-            } catch(Exception ex) {
-                result.completeExceptionally(ex);
             }
-        }).start();
-
-        return result;
+            catch(InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e.getMessage(), e);
+            }
+        }, executorService);
     }
 
     private static String getHostname(final DiscoveryClientConfig config) {
@@ -126,6 +136,10 @@ public class DiscoveryClient implements AutoCloseable {
     public void close() {
         if(httpClientConnectionManager != null) {
             httpClientConnectionManager.close();
+        }
+        if (cleanExecutor == true && executorService != null) {
+            executorService.shutdown();
+            executorService = null;
         }
     }
 }

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
@@ -12,9 +12,6 @@ import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
 
 /**
  * Class for performing network-based discovery of the connectivity properties of registered greengrass cores
@@ -38,18 +35,10 @@ public class DiscoveryClient implements AutoCloseable {
     private final HttpClientConnectionManager httpClientConnectionManager;
 
     /**
-     * We need to use a custom executor to avoid leaving threads alive when the discovery client is closed
-     * It also fixes issues with the SecurityManager as well - as created ExecutorServices created via code
-     * inherit the permissions of the application, unlike the default common thread pool.
-     */
-    private final ExecutorService executorService;
-
-    /**
      *
      * @param config Greengrass discovery client configuration
      */
     public DiscoveryClient(final DiscoveryClientConfig config) {
-        this.executorService = Executors.newFixedThreadPool(1);
         this.httpClientConnectionManager = HttpClientConnectionManager.create(
                 new HttpClientConnectionManagerOptions()
                     .withClientBootstrap(config.getBootstrap())
@@ -70,49 +59,57 @@ public class DiscoveryClient implements AutoCloseable {
         if(thingName == null) {
             throw new IllegalArgumentException("ThingName cannot be null!");
         }
-        return CompletableFuture.supplyAsync(() -> {
-            try(final HttpClientConnection connection = httpClientConnectionManager.acquireConnection().get()) {
-                final String requestHttpPath =  "/greengrass/discover/thing/" + thingName;
-                final HttpHeader[] headers = new HttpHeader[] {
-                        new HttpHeader("host", httpClientConnectionManager.getUri().getHost())
-                };
-                final HttpRequest request = new HttpRequest("GET", requestHttpPath, headers, null);
-                //we are storing everything until we get the entire response
-                final CompletableFuture<Integer> responseComplete = new CompletableFuture<>();
-                final StringBuilder jsonBodyResponseBuilder = new StringBuilder();
-                final Map<String, String> responseInfo = new HashMap<>();
-                try(final HttpStream stream = connection.makeRequest(request, new HttpStreamResponseHandler() {
-                        @Override
-                        public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] httpHeaders) {
-                            Arrays.stream(httpHeaders).forEach(header -> {
-                                responseInfo.put(header.getName(), header.getValue());
-                            });
+
+        CompletableFuture<DiscoverResponse> result = new CompletableFuture<>();
+        // We create and start a thread so it executes asynchronously and independently from the main process. The thread will set the
+        // result CompletableFuture when it is done or if an exception occurs.
+        new Thread(() -> {
+            try {
+                try (final HttpClientConnection connection = httpClientConnectionManager.acquireConnection().get()) {
+                    final String requestHttpPath =  "/greengrass/discover/thing/" + thingName;
+                    final HttpHeader[] headers = new HttpHeader[] {
+                            new HttpHeader("host", httpClientConnectionManager.getUri().getHost())
+                    };
+                    final HttpRequest request = new HttpRequest("GET", requestHttpPath, headers, null);
+                    //we are storing everything until we get the entire response
+                    final CompletableFuture<Integer> responseComplete = new CompletableFuture<>();
+                    final StringBuilder jsonBodyResponseBuilder = new StringBuilder();
+                    final Map<String, String> responseInfo = new HashMap<>();
+
+                    try (final HttpStream stream = connection.makeRequest(request, new HttpStreamResponseHandler() {
+                            @Override
+                            public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] httpHeaders) {
+                                Arrays.stream(httpHeaders).forEach(header -> {
+                                    responseInfo.put(header.getName(), header.getValue());
+                                });
+                            }
+                            @Override
+                            public int onResponseBody(HttpStream stream, byte bodyBytes[]) {
+                                jsonBodyResponseBuilder.append(new String(bodyBytes, StandardCharsets.UTF_8));
+                                return bodyBytes.length;
+                            }
+                            @Override
+                            public void onResponseComplete(HttpStream httpStream, int errorCode) {
+                                responseComplete.complete(errorCode);
+                            }
+                         }))
+                         {
+                            stream.activate();
+                            responseComplete.get();
+                            if (stream.getResponseStatusCode() != 200) {
+                                throw new RuntimeException(String.format("Error %s(%d); RequestId: %s",
+                                        HTTP_HEADER_ERROR_TYPE, stream.getResponseStatusCode(), HTTP_HEADER_REQUEST_ID));
+                            }
+                            final String responseString = jsonBodyResponseBuilder.toString();
+                            result.complete(GSON.fromJson(new StringReader(responseString), DiscoverResponse.class));
                         }
-                        @Override
-                        public int onResponseBody(HttpStream stream, byte bodyBytes[]) {
-                            jsonBodyResponseBuilder.append(new String(bodyBytes, StandardCharsets.UTF_8));
-                            return bodyBytes.length;
-                        }
-                        @Override
-                        public void onResponseComplete(HttpStream httpStream, int errorCode) {
-                            responseComplete.complete(errorCode);
-                        }})) {
-                    stream.activate();
-                    responseComplete.get();
-                    if (stream.getResponseStatusCode() != 200) {
-                        throw new RuntimeException(String.format("Error %s(%d); RequestId: %s",
-                                HTTP_HEADER_ERROR_TYPE, stream.getResponseStatusCode(), HTTP_HEADER_REQUEST_ID));
-                    }
-                    final String responseString = jsonBodyResponseBuilder.toString();
-                    return GSON.fromJson(new StringReader(responseString), DiscoverResponse.class);
-                } catch (InterruptedException | ExecutionException e) {
-                    throw new RuntimeException(e);
                 }
+            } catch(Exception ex) {
+                result.completeExceptionally(ex);
             }
-            catch(InterruptedException | ExecutionException e) {
-                throw new RuntimeException(e.getMessage(), e);
-            }
-        }, executorService);
+        }).start();
+
+        return result;
     }
 
     private static String getHostname(final DiscoveryClientConfig config) {
@@ -129,9 +126,6 @@ public class DiscoveryClient implements AutoCloseable {
     public void close() {
         if(httpClientConnectionManager != null) {
             httpClientConnectionManager.close();
-        }
-        if (executorService != null) {
-            executorService.shutdown();
         }
     }
 }

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
@@ -13,6 +13,8 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 /**
  * Class for performing network-based discovery of the connectivity properties of registered greengrass cores
@@ -36,10 +38,18 @@ public class DiscoveryClient implements AutoCloseable {
     private final HttpClientConnectionManager httpClientConnectionManager;
 
     /**
+     * We need to use a custom executor to avoid leaving threads alive when the discovery client is closed
+     * It also fixes issues with the SecurityManager as well - as created ExecutorServices created via code
+     * inherit the permissions of the application, unlike the default common thread pool.
+     */
+    private final ExecutorService executorService;
+
+    /**
      *
      * @param config Greengrass discovery client configuration
      */
     public DiscoveryClient(final DiscoveryClientConfig config) {
+        this.executorService = Executors.newFixedThreadPool(1);
         this.httpClientConnectionManager = HttpClientConnectionManager.create(
                 new HttpClientConnectionManagerOptions()
                     .withClientBootstrap(config.getBootstrap())
@@ -102,11 +112,11 @@ public class DiscoveryClient implements AutoCloseable {
             catch(InterruptedException | ExecutionException e) {
                 throw new RuntimeException(e.getMessage(), e);
             }
-        });
+        }, executorService);
     }
 
     private static String getHostname(final DiscoveryClientConfig config) {
-        //allow greengrass server endpoint to be manualy set for unique endpoints
+        //allow greengrass server endpoint to be manually set for unique endpoints
         if (config.getGGServerName().equals("")) {
             return String.format("greengrass-ats.iot.%s.%s",
                 config.getRegion(), AWS_DOMAIN_SUFFIX_MAP.getOrDefault(config.getRegion(), AWS_DOMAIN_DEFAULT));
@@ -119,6 +129,9 @@ public class DiscoveryClient implements AutoCloseable {
     public void close() {
         if(httpClientConnectionManager != null) {
             httpClientConnectionManager.close();
+        }
+        if (executorService != null) {
+            executorService.shutdown();
         }
     }
 }

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClient.java
@@ -15,6 +15,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 /**
  * Class for performing network-based discovery of the connectivity properties of registered greengrass cores
@@ -137,9 +138,20 @@ public class DiscoveryClient implements AutoCloseable {
         if(httpClientConnectionManager != null) {
             httpClientConnectionManager.close();
         }
-        if (cleanExecutor == true && executorService != null) {
+        if (cleanExecutor == true) {
             executorService.shutdown();
-            executorService = null;
+            try{
+                // Give the executorService 30 seconds to finish existing tasks. If it takes longer, force it to shutdown
+                if(!executorService.awaitTermination(30,TimeUnit.SECONDS)){
+                    executorService.shutdownNow();
+                }
+            } catch (InterruptedException ie){
+                // If current thread is interrupted, force executorService shutdown
+                executorService.shutdownNow();
+                // Preserve interrupt status
+                Thread.currentThread().interrupt();
+            }
         }
+        executorService = null;
     }
 }

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java
@@ -1,5 +1,7 @@
 package software.amazon.awssdk.iot.discovery;
 
+import java.util.concurrent.ExecutorService;
+
 import software.amazon.awssdk.crt.http.HttpProxyOptions;
 import software.amazon.awssdk.crt.io.ClientBootstrap;
 import software.amazon.awssdk.crt.io.SocketOptions;
@@ -17,6 +19,7 @@ public class DiscoveryClientConfig implements AutoCloseable {
     final private int maxConnections;
     final private HttpProxyOptions proxyOptions;
     final private String ggServerName;
+    private ExecutorService discoveryExecutor;
 
     /**
      * Constructor for DiscoveryClientConfig creates the correct endpoint if not in a special region
@@ -44,6 +47,7 @@ public class DiscoveryClientConfig implements AutoCloseable {
         this.maxConnections = maxConnections;
         this.proxyOptions = proxyOptions;
         this.ggServerName = "";
+        this.discoveryExecutor = null;
     }
 
     /**
@@ -70,7 +74,7 @@ public class DiscoveryClientConfig implements AutoCloseable {
 
     /**
      * Default Constructor for DiscoveryClientConfig that allows the specification of a specific ggServerName to use in special regions
-     * 
+     *
      * @param bootstrap client bootstrap to use to establish network connections
      * @param tlsContextOptions tls configuration for client network connections.  For greengrass discovery, the
      *                          tls context must be initialized with the certificate and private key of the
@@ -94,6 +98,7 @@ public class DiscoveryClientConfig implements AutoCloseable {
         this.maxConnections = maxConnections;
         this.proxyOptions = proxyOptions;
         this.ggServerName = ggServerName;
+        this.discoveryExecutor = null;
     }
 
     /**
@@ -162,6 +167,27 @@ public class DiscoveryClientConfig implements AutoCloseable {
 
     public String getGGServerName() {
         return ggServerName;
+    }
+
+    /**
+     * @return the executor set for this discover client, if one is set.
+     */
+    public ExecutorService getDiscoveryExecutor() {
+        return this.discoveryExecutor;
+    }
+
+    /**
+     * Sets the executor that is used when calling discover().
+     * If set using this function, you are expected to close/shutdown the executor when finished with it.
+     *
+     * If it is not set, the discovery client will internally manage its own executor.
+     *
+     * @param override The executor to use
+     * @return The client config
+     */
+    public DiscoveryClientConfig setDiscoveryExecutor(ExecutorService override) {
+        this.discoveryExecutor = override;
+        return this;
     }
 
     @Override

--- a/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java
+++ b/sdk/src/main/java/software/amazon/awssdk/iot/discovery/DiscoveryClientConfig.java
@@ -170,7 +170,7 @@ public class DiscoveryClientConfig implements AutoCloseable {
     }
 
     /**
-     * @return the executor set for this discover client, if one is set.
+     * @return the ExecutorService set for this discover client config. Returns null if one hasn't been set.
      */
     public ExecutorService getDiscoveryExecutor() {
         return this.discoveryExecutor;

